oeps/architectural-decisions/oep-0065-frontend-composability.rst

```diff
@@ -0,0 +1,492 @@
+
+User and Developer Experience
+=============================
+
+Bundle Size
+-----------
+
+Each MFE bundles all of its own dependencies, even if they're the same version as another MFE. This means that as a user navigates between MFEs, they end up downloading the same set of common dependencies over and over again. This can result in megabytes of overhead for the average user navigating between a few MFEs, and slows down the entire experience.
+
+Full Page Refreshes
+-------------------
+
+Each MFE has its own index.html page, and needs to load all its own dependencies from scratch whenever you navigate to it. This means that the browser performs a full page refresh each time a user transfers from MFE to MFE.
+
+Build time
+----------
+
+Each MFE must bundle all its own dependencies at build time, creating significant overhead for the Webpack build process. When building multiple MFEs, this additional, repeated overhead adds up quickly, making builds prohibitively slow.
+
+Dependency Maintenance
+----------------------
+
+Since each MFE has its own complete set of dependencies, the overhead of keeping them all up to date can be overwhelming. Security patches, bug fixes, new features, and breaking changes all add up and create a significant maintenance burden.
+
+Composability
+=============
+
+We have no clean, low-overhead way of composing components from multiple MFEs into a single page, or extending an MFE with additional functionality.
+
+The reality of MFEs is that while each application may represent a cohesive domain or bounded context, sometimes you need content and functionality from more than one domain on the page at the same time.
+
+Alternately, providers may want to show different *versions* of MFEs to different users while keeping the rest of the app (header, navigation, other MFEs) unchanged.
+
+We aren't without options here, but what we have today has its own issues. It's worth walking through what we have today.
+
+Shared Libraries (Current Option #1)
+------------------------------------
+
+Because each MFE is siloed from each other - both in repositories and at runtime - our only viable mechanism for sharing code has historically been to extract it into a shared library that gets bundled with each at build time. This creates more repository/dependency overhead.
+
+Build-time package overrides (Current Option #2)
+------------------------------------------------
+
+NPM and package.json allow operators to override dependency resolution by installing an alternate version of a dependency prior to build time. This has historically been how we've allowed operators to override the header, footer, and brand.
+
+The system is confusing, somewhat brittle, and only works at build time. If an operator needs different headers/footers/brands for different clients, we've just multiplied the number of builds they need to do (already a problem even without this - see the "Build time" issue above!)
```

Comment:
```suggestion
The system is confusing, somewhat brittle, and only works at build time. If a site operator needs different headers/footers/brands for different clients, we've just multiplied the number of builds they need to do (already a problem even without this - see the "Build time" issue above!)
```
Is "client" the right word to use? That describes a _way of running your business_ rather than a way of running your site. What's the technical term for what a site operator is doing when they desire to do this? Sorry I clearly don't devops

Reply:
Maybe "if a site operator is managing multiple instances with custom header/footer/brand requirements, we've just multiplied..."

Reply:
We use `open edx instance` or just `instance` to refer to what I think is meant by client here. A client as part of a business might even have many instances.

Reply:
As discussed in chat with David, another possible term to incorporate here could be "multi-tenant"/"multi-tenancy"

Reply:
If that is a goal it should definitely be explicitly stated!

Reply:
I had a conversation with @felipemontoya the other day where he encouraged me _not_ to worry about multi-tenancy as a core goal of this. 😄  I think we can simplify this by saying "instances" instead of "clients" as he suggests, which removes the "business" language.